### PR TITLE
[#548] Update Skills tests' base structure

### PR DESCRIPTION
**Pull request by sw-joelmut**
Addresses # 548

**Note:** This PR depends on the changes in the PRs # 573 and # 574.

## Description
This PR refactors the `Skills` test structure, separating what needs to be in the `Common` folder, so `Standalone` bots can use the functionalities, additionally reduces duplicated functionality and separates the `CardActions` scenario into multiple classes to speed up the process, it also adds the new `Authentication` tests.

### Detailed Changes
- Adds a new `TestCase` base class with common properties when building the test cases.
- Updates and moves the `TestCaseDataObject` to the Common folder.
- Updates the `ScriptTestBase` including the `Channels` list, since it's common across bots (Skills, Standalone).
- Updates all `CardActions` tests separating the scripts into its own test class, reducing the time it takes to execute this scenario locally or in the pipelines.
- Adds a new `CardTestBase` containing the base test and exclusions needed across `Cards'` test classes.
- Updates the Skill list from `Class` to an `Enum`, being consistent like its Host counterpart.
- Adds a new `SkillsTestCase` class containing all the lists (DeliveryModes, Hosts, Skills, etc) necessary to make the test cases combinations, as well as the `Cardinal Product` method to construct the test cases.
- Updated all tests to inherit from the `SkillsTestCase` class, reducing the amount of duplicated code required to make the test combinations.
- Adds new `Authentication` tests and bots, to test against `UserAssignedMSI` and `SingleTenant` for DotNet and JS.

## Testing
The following images show the Test Scenarios pipeline continue working as expected, and a sneak peek of how the new changes look compared to the previous version.
imagen
imagen

### CODE REVIEW

File: Tests/Functional/Skills/CardActions/CardBaseTests.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        public CardBaseTests(ITestOutputHelper output)
+            : base(output)
+        {
+        }
+
+        public static Func<SkillsTestCase, bool> Exclude(Func<SkillsTestCase, bool> exclude = default)
+        {
+            // TODO: Enable after fixing differences in the Cards scripts.
+            return (SkillsTestCase test) => test.Skill == SkillBot.ComposerSkillBotDotNet || (exclude != null && exclude(test));
+        }
```
Comment (ceciliaavila) [inline Tests/Functional/Skills/CardActions/CardBaseTests.cs]: This won't be needed if the PR enabling cards get merged first.

File: Tests/Functional/Skills/Common/SkillBot.cs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        /// </summary>
+        EchoSkillBotComposerDotNet,
+
+        /// <summary>
+        /// Echo MultiTenant skill implemented using using DotNet 3.1.
+        /// </summary>
+        EchoSkillBotDotNet,
+
+        /// <summary>
+        /// Echo UserAssignedMSI skill implemented using using DotNet 3.1.
```
Comment (ceciliaavila) [inline Tests/Functional/Skills/Common/SkillBot.cs]:
```suggestion
        /// Echo UserAssignedMSI skill implemented using DotNet 3.1.
```

File: Tests/Functional/Skills/Common/SkillBot.cs [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+{
+    public enum SkillBot
+    {
+        /// <summary>
+        /// Echo MultiTenant skill implemented using Composer and the DotNet runtime.
+        /// </summary>
+        EchoSkillBotComposerDotNet,
+
+        /// <summary>
+        /// Echo MultiTenant skill implemented using using DotNet 3.1.
```
Comment (ceciliaavila) [inline Tests/Functional/Skills/Common/SkillBot.cs]:
```suggestion
        /// Echo MultiTenant skill implemented using DotNet 3.1.
```

File: Tests/Functional/Skills/Common/SkillBot.cs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        /// </summary>
+        EchoSkillBotDotNet,
+
+        /// <summary>
+        /// Echo UserAssignedMSI skill implemented using using DotNet 3.1.
+        /// </summary>
+        EchoSkillBotDotNetMSI,
+
+        /// <summary>
+        /// Echo SingleTenant skill implemented using using DotNet 3.1.
```
Comment (ceciliaavila) [inline Tests/Functional/Skills/Common/SkillBot.cs]:
```suggestion
        /// Echo SingleTenant skill implemented using DotNet 3.1.
```

File: Tests/Functional/Skills/Common/SkillBot.cs [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        /// </summary>
+        EchoSkillBotDotNetST,
+
+        /// <summary>
+        /// Echo MultiTenant v3 skill implemented using using DotNet 3.1.
```
Comment (ceciliaavila) [inline Tests/Functional/Skills/Common/SkillBot.cs]: The same in the rest of the comments.

File: Tests/Functional/Skills/SignIn/SignInTests.cs [hunk 6]
```diff
@@ -18,71 +18,40 @@
...
-        public static IEnumerable<object[]> TestCases()
-        {
-            var channelIds = new List<string> { Channels.Directline };
-            var deliverModes = new List<string>
-            {
-                DeliveryModes.Normal,
-                DeliveryModes.ExpectReplies,
-            };
+        // TODO: Enable after fixing issue (The test is failing with timeout after the signIn).
+        public static bool Exclude(SkillsTestCase test) => test.Skill == SkillBot.ComposerSkillBotDotNet;
```
Comment (ceciliaavila) [inline Tests/Functional/Skills/SignIn/SignInTests.cs]: This won't be necessary if the PR enabling signIn gets merged first.

**Review by ceciliaavila — CHANGES_REQUESTED** [review]

**Comment by sw-joelmut** [comment]
@ceciliaavila fixed, thanks!

**Review by ceciliaavila — APPROVED** [review]

**Pull request closed without merging**